authors/apps/authentication/serializers.py

```diff
@@ -1,34 +1,113 @@
+import re
+
 from django.contrib.auth import authenticate
 
+from django.core.validators import RegexValidator
 from rest_framework import serializers
+from rest_framework.validators import UniqueValidator
 
 from .models import User
 
+email_expression = re.compile(
+    r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")
+at_least_number = re.compile(
+    r"^(?=.*[0-9]).*")
+at_least_uppercase = re.compile(
+    r"^(?=.*[A-Z])(?=.*[a-z])(?!.*\s).*")
+at_least_special_char = re.compile(
+    r"^[a-zA-Z0-9_]*$")
 
 class RegistrationSerializer(serializers.ModelSerializer):
-    """Serializers registration requests and creates a new user."""
-
+    # Username must be longer than 4 characters but shorter than 128 characters. 
+    # Usernamr is a required field
+    username = serializers.CharField(
+        max_length=128,
+        min_length=4,
+        required=True,
+        error_messages={
+            "required"  : "Username is required!"
+        }
+    )
+    # Ensure email is present and has the valid format example@mail.com
+    email = serializers.EmailField(
+        required=True,
+        error_messages={
+            "required" : "Email is required!"
+        }
+    )
     # Ensure passwords are at least 8 characters long, no longer than 128
     # characters, and can not be read by the client.
-    password = serializers.CharField(
+    password = serializers.RegexField(
+        regex="^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$",
         max_length=128,
         min_length=8,
-        write_only=True
+        write_only=True,
+        required=True,
+        error_messages={
+            "required"  : "Password is required!",
+            "invalid"   : "Password MUST have BOTH letters and numbers",
+            "invalid"   : "Password MUST be longer than eight(8) charcaters!",
+            "invalid"   : "Password is too long (max=128 characters)"
+        }
     )
-
     # The client should not be able to send a token along with a registration
     # request. Making `token` read-only handles that for us.
+    token = serializers.CharField(
+        read_only=True
+    )
+    
+    def create(self, validated_data):
+        # Use the `create_user` method we wrote earlier to create a new user.
+        return User.objects.create_user(**validated_data)
+
+    def validate(self, data):
```

Comment:
This method works but as **Uncle Bob** says in his book on clean architecture, methods should be small. you can resolve this by using **Field-level validation** instead of this. In this case, you can validate every field on its own method.

Reply:
For the identified discrepancy on the README.md, the current version of the file living on the develop branch has two badges for builds and coverage. I believe that is why it should stay to avoid merge conflicts. Finally, I modularised the previously overloaded `validate` function to separate field-based validations: `validate_username` , `validate_email` , and `validate_password`. These changes are reflected in the new branch I created `ft-signup-validators-160577607`